Constants overhaul

daboross opened the pull request:
This implements the ability to run `look` and `find` on power creeps, adds some power creep related constants, and adds other misc. constant updates.

This does add a `PowerCreep` type, but it doesn't add any implementation to it beyond it being a `RoomObject`. This should be enough to at least notice power creeps existing, but not really enough to do anything with one's own power creeps.

Fixes #58.

ASalvail commented on screeps-game-api/javascript/utils.js:
```diff
@@ -89,7 +90,7 @@ function __structure_type_str_to_num(str) {
         case STRUCTURE_CONTAINER: return 17;
         case STRUCTURE_NUKER: return 18;
-        default: throw new Error("unknown resource type " + str);
+        default: throw new Error("unknown structure type " + str);
```
Haha, good catch!

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -979,3 +995,60 @@ impl TryFrom<Value> for ResourceType {
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "constants-serde", derive(Serialize, Deserialize))]
+pub enum PowerClass {
+    #[cfg_attr(feature = "constants-serde", serde(rename = "operator"))]
```
You just removed those higher up. Can this one also be changed to `#[serde(rename(...)]`?

daboross replied:
Hm - I mean we could! I removed the ones above because we used them within the codebase, while these are still feature gated so we don't generate extra code. It is somewhat arbitrary that serde implementations are feature gated this way but nothing else is, though.

Maybe we should remove these gates and just have everything implement it all the time!

ASalvail replied:
Do you remember why we wanted to feature gate them? While we might save some space, this is something that would trip newcomers.

daboross replied:
I don't really - I think I just personally needed them and didn't want to bloat the binary?

I'll remove the feature gate - they're useful enough.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -979,3 +995,60 @@ impl TryFrom<Value> for ResourceType {
+    OperateFactory = 19,
+}
+
+// TODO: POWER_INFO
```
Wasn't there anything about tunnels also? I'm uncomfortable bumping the date at the top of the file without catching everything that was updated in the constants.
Also, how hard could it be to make a macro that uses the `js` file that describes the constants directly to generate this file?

daboross replied:
I think the only tunnel thing was `CONSTRUCTION_COST_ROAD_WALL_RATIO`, which I added above.

But I agree with the concerns here - I'll either implement the rest of them or make a summary TODO and TODO issue.

daboross replied:
With automated stuff, I bet we could do something for all the "plain" constants which are just numbers. But I think it might be hard translating to enums like ResourceType, PowerType, and such. I mean I haven't even figured out how we want to translate reactions nor power info - I _think_ functions which match on arguments and return the result would be the best way, but that might not be true.

ASalvail replied:
Maybe later...

ASalvail commented on screeps-game-api/src/lib.rs:
```diff
@@ -29,6 +29,7 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 extern crate serde_json;
+extern crate serde_repr;
```
You'll have to rebase off of the 2018 rustification 😞

daboross commented:
Rebased on master - going to add more constants though to see how close I can get to completion.

daboross commented:
Rebased after #166 and did a lot more constant management.

Added some breaking changes which may or may not want to be part of this PR?

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -16,6 +16,7 @@
 
 use log::error;
 use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
```
Isn't that shadowing the previous import?

daboross replied:
One's the derive and the other's the actual trait for usage - I think both are necessary?

ASalvail replied:
TIL that macros and items are not in the same namespace. Interesting!

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -451,30 +447,12 @@ impl Part {
-        Ok(Self::try_from(x)
-            .unwrap_or_else(|()| panic!("JavaScript gave unknown part constant {}", x)))
+        Ok(Self::from_u32(x)
+            .unwrap_or_else(|| panic!("JavaScript gave unknown part constant {}", x)))
```
Shouldn't you return an `Err`?

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -655,28 +633,7 @@ impl TryFrom<Value> for StructureType {
-            18 => StructureType::Nuker,
-            _ => panic!("unknown structure type integer {}", x),
-        })
+        Ok(Self::from_u32(x).unwrap_or_else(|| panic!("unknown structure type integer {}", x)))
```
Same with `Err`.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -1115,52 +1070,8 @@ impl TryFrom<Value> for ResourceType {
-            _ => panic!("unknown resource type integer {}", x),
-        })
+
+        Ok(Self::from_u32(x).unwrap_or_else(|| panic!("unknown resource type integer {}", x)))
```
`Err`

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -1217,3 +1129,13 @@ pub enum PowerType {
+    fn try_from(v: Value) -> Result<Self, Self::Error> {
+        let x: u32 = v.try_into()?;
+
+        Ok(Self::from_u32(x).unwrap_or_else(|| panic!("unknown power type integer {}", x)))
```
`Err`

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -14,13 +14,12 @@
 use serde::{Deserialize, Serialize};
 use serde_repr::{Deserialize_repr, Serialize_repr};
-use stdweb::{Number, Reference, Value};
+use stdweb::{Number, Reference, Value, __js_deserializable_serde_boilerplate, js_deserializable};
```
if you want to avoid weird macro imports all over the place, feel free to use `crate::macros::*` which exposes those macros.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -233,17 +217,7 @@ pub enum Direction {
-        }))
-    }
-}
+js_deserializable!(Direction);
```
I thought we had agreed not to use `js_deserializable!` since it was too costly?

daboross replied:
I think it should be better with `stdweb` 0.4.9+! Deserializing now no longer uses json parsing, and instead implements `TryFrom<Value>` by traversing the value with serde, calling into javascript to get any references.

I think it should be fairly equivalent to a manual implementation of `TryFrom` now, especially with these integer constants which should just be represented by `Value::Number`.

ASalvail replied:
Oh, that's great news, I didn't know!

daboross replied:
it is! I should have said something though rather than just barging ahead.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -513,7 +517,7 @@ pub const STORAGE_HITS: u32 = 10_000;
 /// Translates `STRUCTURE_*` constants.
 #[repr(u32)]
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, FromPrimitive, Serialize, Deserialize)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
```
I'm not sure I understand the difference between `Exit` which uses `FromPrimitive` and `StructureType` which doesn't. They're both integers. Unless you want to be pedantic and say that `u32` are natural numbers and not integers 😛

daboross replied:
Main difference is that in screeps `StructureType` is actually represented by strings like `"spawn"`, `"extension"`, `"road"`, etc. and the numbering code we use is entirely internal to `screeps-game-api`. I mean it uses `__structure_type_num_to_str` from `javascript/utils.js` to translate integers to the actual constant names.

I was thinking it'd be nice to have `FromPrimitive` for screeps constants which actually are integer constants, like Exit, Direction, ReturnCode, Find*, Color since users might actually encounter those? But it's mostly contrived reasoning to find some way to make it reasonable to have `FromPrimitive` on `ReturnCode` in particular since we use that.

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -896,47 +896,82 @@ pub enum ResourceType {
             Energy | Power | Hydrogen | Oxygen | Utrium | Lemergium | Keanium | Zynthium
-            | Catalyst | Ghodium | Ops => 0,
+            | Catalyst | Ops => return None,
+            // OH: 20,
```
Yeah! That and to double check I actually have the right values and to make it easier to compare to future `constants.js` files.

ASalvail replied:
Ah, you copy pasted the values, commented them then inserted the variants. It makes more sense now.

daboross replied:
Yep!

daboross commented:
Thanks for the review! I have a lot to fix up and add comments for.

> I'm concerned with the very uneven treatment of TryFrom implementations

With this, I think it should be fairly consistent in that everything implements `TryFrom<Value>`, but I agree that the "extra" impls are kind of scattered around... I left in the ones for `ReturnCode` and `Exit` mostly because I didn't want to touch more code outside of the constants file, but I'll go in and remove those now.

ASalvail commented:
By uneven treatment, I mostly referred to the `try_from` implementations that `panic!` rather than return an error. I get that it should probably never happen, but it also sounds like the exact reason we're doing `try_from` rather than `from`.

ASalvail commented:
Fix #58

ASalvail commented:
Addresses some of #88

daboross commented on screeps-game-api/javascript/utils.js:
```diff
@@ -188,6 +190,7 @@ function __resource_type_str_to_num(str) {
         case RESOURCE_CATALYZED_ZYNTHIUM_ALKALIDE: return 41;
         case RESOURCE_CATALYZED_GHODIUM_ACID: return 42;
         case RESOURCE_CATALYZED_GHODIUM_ALKALIDE: return 43;
+        case RESOURCE_OPS: return 44;
```
I think it has something to do with power creeps? There's a power named `PWR_GENERATE_OPS` which seems like it could be related.

I found the constant `RESOURCE_OPS` in the constants file at https://github.com/screeps/common/blob/068343753adf090fd1991944d2292be9e236b7dd/lib/constants.js#L406, and it is counted as part of "all resources" like the other ones at https://github.com/screeps/common/blob/068343753adf090fd1991944d2292be9e236b7dd/lib/constants.js#L989.

daboross replied:
It does seem to be used in engine: at https://github.com/screeps/engine/blob/a4e431c9ee33dcc55c04f3adc479145c210c73fd/src/processor/intents/power-creeps/usePower.js#L58.

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -732,6 +799,33 @@ impl TryFrom<Value> for Density {
+    /// Translates the `MINERAL_DENSITY_PROBABILITY` constant.
+    ///
+    /// All values are between 0 and 1, but the total is roughly `2.5` so these
+    /// aren't percentages.
```
Ah - looks like this is right!

https://github.com/screeps/engine/blob/a4e431c9ee33dcc55c04f3adc479145c210c73fd/src/processor/intents/minerals/tick.js#L24-L29

daboross commented:
Re-rebased on master, and fixed up!

In addition to addressing the comments made here, it also includes some new `find` module improvements, and a small `Density` improvement.

As a note, I think there's _something_ that this PR breaks, since I've been running my AI using it and it's been panicking on a find call somewhere. I'm not sure what exactly is broken, but once I find it I'll make a note of it here.

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -199,6 +209,13 @@ pub mod find {
+        EXIT_TOP, Exit::Top as i32, RoomPosition;
+        EXIT_RIGHT, Exit::Right as i32, RoomPosition;
+        EXIT_BOTTOM, Exit::Bottom as i32, RoomPosition;
+        EXIT, Exit::All as i32, RoomPosition;
```
These should help with the situation we last talked about in the slack chat!

I might be putting too much in this PR. But what else is an "overhaul" for?

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -656,7 +681,31 @@ pub fn controller_levels(current_rcl: u32) -> u32 {
-// TODO: controller_*
+/// Translates the `CONTROLLER_DOWNGRADE` constant.
+///
+/// Accepts levels 1-7.
```
This is the only thing that wasn't fixed that I can see.

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -172,6 +211,10 @@ pub mod find {
         MY_POWER_CREEPS, 120, PowerCreep;
         HOSTILE_POWER_CREEPS, 121, PowerCreep;
+        EXIT_TOP, Exit::Top as i32, RoomPosition;
+        EXIT_RIGHT, Exit::Right as i32, RoomPosition;
```
thank you

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -59,54 +61,91 @@ impl ReturnCode {
+    /// all-upper-case constants in [this module](crate::constants::find) will
+    /// be more helpful, and won't require casting the result types.
+    ///
+    /// If you do use this constant, I recommend just importing the `find`
```
That sounds better, yeah. I wanted to write something "softer" than that, but since code really does look weird without `find::` beforehand "should" does seem good.

daboross replied:
In other places I seem to use "it's recommended", which is slightly better. But _should_ in this case seems right, still.

ASalvail replied:
_You may want_ ?

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -658,7 +659,7 @@ pub const CONSTRUCTION_COST_ROAD_WALL_RATIO: u32 = 150;
 /// Translates the `CONTROLLER_LEVELS` constant.
 ///
-/// Accepts levels 1-7.
+/// Accepts levels 1-8.
```
Argh, no, this one is actually 1-7 😆

daboross replied:
:man_facepalming:

daboross commented:
While I'm super excited for this, there is one problem: I assumed that `#[derive(Deserialize)]` in combination with serde-json and/or stdweb's Value deserializer implementation, enum variants could be deserialized from both integers and strings.

For these formats, when deserializing an enum, only string names for the variants are supported.

daboross commented:
Required adding a bit more code, but decoding all the constants which are encoded as integers but are actually strings in game should work now.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -448,8 +458,16 @@ pub mod look {
+/// *Note:* This constant's `TryFrom<Value>`, `Serialize` and `Deserialize`
+/// implementations only operate on made-up integer constants. If you're ever
+/// using these impls manually, you need to use the `__part_num_to_str` and
+/// `__part_str_to_num` JavaScript functions to convert.
```
hopefully you can copy paste the previous example...
Or maybe make one at the module level and refer to it here?

daboross replied:
Module-level doc sounds like a good idea.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -361,8 +361,8 @@ js_deserializable!(Color);
 /// _only work with the integer constants_. If you're ever consuming strings
-/// such as `"plain"`, `"swamp"`, `"wall"`, you must convert those manually into
-/// `0`, `1` and `2` respectively.
+/// such as `"plain"`, `"swamp"`, `"wall"`, you can use the
```
Only in JS though. Might be worthwhile to make a `TryFrom<String>` equivalent?

daboross replied:
Added! I gave it a `FromStr` implementation like some of the other constants.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -390,39 +394,53 @@ js_deserializable!(Terrain);
 /// using these impls manually, use the `__look_num_to_str` and
 /// `__look_str_to_num` JavaScript functions to convert.
+///
+/// To convert from a string representation directly, use
```
As mentioned elsewhere, maybe specify that the best method depends on where the value lives?
Actually, it might be beneficial (in another PR) to add a brief explanation of the memory model in the crate's doc. It would avoid some confusion and explain design decisions. Thoughts?

daboross replied:
Adding documentation on the memory model in general sounds like a good idea!

For this particular PR, I've mentioned the two different contexts where one could use the two different conversion methods in the longer documentation for the module.

ASalvail commented on screeps-game-api/src/constants.rs:
```diff
@@ -390,39 +394,53 @@ js_deserializable!(Terrain);
+    /// Helper function for deserializing from a string rather than a fake
+    /// integer value.
+    pub fn deserialize_from_str<'de, D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let s: Cow<'de, str> = Cow::deserialize(d)?;
```
I really don't get the usage of the `Cow`? Serde doesn't allocate memory for strings? Is it because it may still refers to the initial "text" that is deserialized?

daboross replied:
`serde` won't do any allocations, but the deserializer implementations may or may not need to. `Cow` is just a way of encoding that this code is OK with taking either an allocated owned string, or a borrowed one.

For instance, when using `serde_json::from_str` to deserialize, strings without escape patterns can be deserialized as borrowing from the original string passed in. But in `serde_json::from_value`, the ownership of strings is passed from the value into the deserialized data, so we need to be able to accept an owned string too.

It'd function if this code just used `s: String`, but then it would also allocate slightly more than necessary in some situations.

daboross commented on screeps-game-api/src/constants.rs:
```diff
@@ -380,8 +385,13 @@ js_deserializable!(Terrain);
+/// *Note:* This constant's `TryFrom<Value>`, `Serialize` and `Deserialize`
+/// implementations only operate on made-up integer constants. If you're ever
+/// using these impls manually, you need to use the `__look_num_to_str` and
+/// `__look_str_to_num` JavaScript functions to convert.
```
Added an example to module-level doc.
